sql: internal error: no output column equivalent

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5238654&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5238654&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on release-22.1 @ [f143fd81c898f540a0a0db467b5f18beb32644cf](https://github.com/cockroachdb/cockroach/commits/f143fd81c898f540a0a0db467b5f18beb32644cf):


```
The test failed on branch=release-22.1, cloud=gce:
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:304,test_runner.go:875: error: pq: internal error: no output column equivalent to 5
		stmt:
		SELECT
			tab_30745.col3_11 AS col_65710,
			NULL AS col_65711,
			NULL AS col_65712,
			e'\x13L|9\x03<\f':::STRING AS col_65713,
			tab_30745.col3_9 AS col_65714,
			true AS col_65715,
			parse_time(('28.81.114.123/16':::INET::INET || tab_30745.col3_1::STRING)::STRING::STRING)::TIME AS col_65716,
			tab_30745.col3_1 AS col_65717,
			tab_30745.col3_10 AS col_65718
		FROM
			defaultdb.public.table3@table3_col3_12_col3_9_expr_col3_0_expr1_col3_13_expr2_col3_10_col3_14_col3_16_col3_4_key
				AS tab_30745
		ORDER BY
			tab_30745.col3_1 ASC, tab_30745.col3_2 ASC, tab_30745.col3_12 ASC
		LIMIT
			2:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #81632 roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed [C-test-failure O-roachtest O-robot T-sql-queries branch-release-21.2 release-blocker]
- #81356 roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed [C-test-failure O-roachtest O-robot T-sql-queries branch-master]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-16009

--- a/pkg/sql/opt/ordering/lookup_join.go
+++ b/pkg/sql/opt/ordering/lookup_join.go
@@ -82,7 +82,13 @@ func indexJoinBuildProvided(expr memo.RelExpr, required *props.OrderingChoice) o
 	// using column equivalencies.
 	indexJoin := expr.(*memo.IndexJoinExpr)
 	rel := indexJoin.Relational()
-	return remapProvided(indexJoin.Input.ProvidedPhysical().Ordering, &rel.FuncDeps, rel.OutputCols)
+	input := indexJoin.Input
+	// The index join's FDs may not include all the necessary columns for
+	// remapping, so we add the input's FDs as well. See `buildIndexJoinProps`.
+	var fds props.FuncDepSet
+	fds.CopyFrom(&input.Relational().FuncDeps)
+	fds.AddFrom(&rel.FuncDeps)
+	return remapProvided(input.ProvidedPhysical().Ordering, &fds, rel.OutputCols)
 }
 
 func lookupJoinBuildProvided(expr memo.RelExpr, required *props.OrderingChoice) opt.Ordering {

--- a/pkg/sql/opt/xform/testdata/rules/join
+++ b/pkg/sql/opt/xform/testdata/rules/join
@@ -12378,3 +12378,37 @@ project
       │    └── columns: col0:8
       └── filters
            └── col3:4 = col0:8 [outer=(4,8), constraints=(/4: (/NULL - ]; /8: (/NULL - ]), fd=(4)==(8), (8)==(4)]
+
+# Regression for #81649
+exec-ddl
+CREATE TABLE t81649 (
+  col0 STRING NULL,
+  col1 DECIMAL NOT NULL,
+  col2 STRING NULL AS (lower(col0)) STORED,
+  col3 DECIMAL NOT NULL AS (col1 + 0:::DECIMAL) VIRTUAL,
+  PRIMARY KEY (col3, col1),
+  UNIQUE (col3 ASC, col1 DESC) STORING (col0));
+----
+
+opt
+SELECT col2 FROM t81649@t81649_col3_col1_key ORDER BY col0 ASC, col3 ASC LIMIT 2;
+----
+index-join t81649
+ ├── columns: col2:3  [hidden: col0:1 col3:4!null]
+ ├── cardinality: [0 - 2]
+ ├── key: (4)
+ ├── fd: (1)-->(3), (4)-->(1,3)
+ ├── ordering: +1,+4
+ └── top-k
+      ├── columns: col0:1 col1:2!null col3:4!null
+      ├── internal-ordering: +1,+(2|4)
+      ├── k: 2
+      ├── cardinality: [0 - 2]
+      ├── key: (4)
+      ├── fd: (2,4)-->(1), (2)==(4), (4)==(2)
+      ├── ordering: +1,+(2|4) [actual: +1,+2]
+      └── scan t81649@t81649_col3_col1_key
+           ├── columns: col0:1 col1:2!null col3:4!null
+           ├── flags: force-index=t81649_col3_col1_key
+           ├── key: (4)
+           └── fd: (2,4)-->(1), (2)==(4), (4)==(2)

